File: src/input/core/hotkeys.c
#include <assert.h>
#include <stddef.h>   /* NULL */
#include "hotkeys.h"

/* A faithful port of ../../input/hotkeys.py -- combo edge detection + output masking. */

void hotkeys_init(hotkeys_t *h)
{
    assert(h != NULL);
    h->count = 0;
    h->prev = 0;
}

int32_t hotkeys_add(hotkeys_t *h, buttons_t combo, hotkey_action_t action, uint8_t param)
{
    assert(h != NULL);
    assert(combo != 0);                            /* a hotkey needs at least one button */
    if (h->count >= HOTKEY_MAX)
        return -1;
    int32_t i = (int32_t)h->count++;
    h->hotkeys[i].combo = combo;
    h->hotkeys[i].action = action;
    h->hotkeys[i].param = param;
    return i;
}

buttons_t hotkeys_step(hotkeys_t *h, buttons_t pressed, hotkey_fire_t *fired,
                       uint32_t cap, uint32_t *nfired)
{
    assert(h != NULL);
    assert(nfired != NULL);

    buttons_t out = pressed;
    uint32_t n = 0;
    for (uint32_t i = 0; i < h->count; i++) {
        buttons_t combo = h->hotkeys[i].combo;
        if ((pressed & combo) == combo) {          /* combo fully held */
            out &= ~combo;                          /* mask it from the output */
            if ((h->prev & combo) != combo) {       /* edge: not fully held last tick */
                if (fired != NULL && n < cap) {
                    fired[n].action = h->hotkeys[i].action;
                    fired[n].param = h->hotkeys[i].param;
                    n++;
                }
            }
        }
    }
    h->prev = pressed;
    *nfired = n;
    /* negative space: masking only removes buttons, it never invents one */
    assert((out & ~pressed) == 0);
    return out;
}

File: test/reference/input/hotkeys.py
"""
Hotkeys -- button COMBOS that fire config actions instead of reaching the game.

A hotkey is a set of buttons (a combo). It FIRES on the tick the combo becomes fully
pressed (an edge -- all its buttons down now, not all down last tick), so holding it
fires once. While the combo is held its buttons are MASKED from the output (it's a
command, not gameplay -- e.g. SELECT+DOWN switches SOCD without sending SELECT+DOWN).

Runs late in the pipeline on resolved logical buttons. Actions are opaque codes here
(see ACTIONS); APPLYING them (profile switch, SOCD cycle, ...) is the config/profiles
layer's job, not this module's. Stateless except the one-tick edge memory.
"""
# Action codes -- kept in lockstep with hotkey_action_t in firmware/core/hotkeys.h.
ACTIONS = ("none", "profile_next", "profile_prev", "profile_set",
           "socd_cycle", "socd_set", "sync_toggle", "turbo_up", "turbo_down",
           "dpad_mode", "fourway_toggle", "reboot_bootloader", "webconfig")


class Hotkeys:
    def __init__(self, hotkeys=()):
        # hotkeys: iterable of (combo_button_names, action_code, param)
        self.hotkeys = [(frozenset(c), a, p) for c, a, p in hotkeys]
        self._prev = frozenset()

    def step(self, pressed):
        """(pressed names) -> (masked output names, [(action, param), ...] fired)."""
        pressed = set(pressed)
        out = set(pressed)
        fired = []
        for combo, action, param in self.hotkeys:
            if combo <= pressed:                    # combo fully held
                out -= combo                        # mask from output
                if not (combo <= self._prev):       # edge: wasn't fully held last tick
                    fired.append((action, param))
        self._prev = frozenset(pressed)
        return out, fired
